File: source/ml/cnn/cnn.cpp
/**
 * @brief Convolutional neural network (CNN) implementation details.
 */
#include <algorithm>
#include <iostream>

#include "ml/cnn/cnn.h"
#include "ml/factory/interface.h"
#include "ml/types.h"
#include "ml/utils.h"

namespace ml::cnn
{
// -----------------------------------------------------------------------------
Cnn::Cnn(factory::Interface& factory, const std::size_t convInput, const std::size_t convKernel, 
         const act_func::Type convFunc, const std::size_t poolSize, 
         const std::size_t denseOutput, const act_func::Type denseFunc)
    : myConvLayers{}
    , myDenseLayers{}
    , myFlattenLayer{nullptr}
    , myFactory{factory}
{
    // Initialize the convolutional layers.
    myConvLayers.emplace_back(factory.convLayer(convInput, convKernel, convFunc));
    myConvLayers.emplace_back(factory.maxPoolLayer(convOutputSize(), poolSize));

    // Initialize the flatten layer.
    myFlattenLayer = factory.flattenLayer(convOutputSize());

    // Initialize the dense layer.
    const std::size_t denseInput{myFlattenLayer->outputSize()};
    myDenseLayers.emplace_back(factory.denseLayer(denseInput, denseOutput, denseFunc));
}

// -----------------------------------------------------------------------------
Cnn::~Cnn() noexcept = default;

// -----------------------------------------------------------------------------
std::size_t Cnn::inputSize() const noexcept 
{ 
    return myConvLayers[0U]->inputSize(); 
}

// -----------------------------------------------------------------------------
std::size_t Cnn::outputSize() const noexcept 
{
    const std::size_t last{myDenseLayers.size() - 1U};
    return myDenseLayers[last]->outputSize();
}

// -----------------------------------------------------------------------------
const Matrix1d& Cnn::predict(const Matrix2d& input) noexcept 
{
    feedforward(input);
    return output();
}

// -----------------------------------------------------------------------------
void Cnn::addDenseLayer(const std::size_t outputSize, const act_func::Type actFunc)
{
    myDenseLayers.emplace_back(myFactory.denseLayer(this->outputSize(), outputSize, actFunc));
}

// -----------------------------------------------------------------------------
bool Cnn::train(const Matrix3d& trainIn, const Matrix2d& trainOut, const std::size_t epochCount,
                const double learningRate)
{
    if (0.0 >= learningRate)
    {
        std::cerr << "Failed to train CNN: invalid learning rate " << learningRate << "!\n";   
    }
    else if (0U == epochCount)
    {
        std::cerr << "Failed to train CNN: invalid epoch count " << epochCount << "!\n";   
    }

    const std::size_t setCount{std::min(trainIn.size(), trainOut.size())};

    if (0U == setCount)
    {
        std::cerr << "Failed to train CNN: invalid set count " << setCount << "!\n";
    }

    // Create a training order list.
    TrainOrderList trainOrder{createTrainOrderList(setCount)};

    // Train the network the specified number of epochs.
    for (std::size_t i{}; i < epochCount; ++i)
    {
        // Shuffle the training order list at the start of each epoch.
        shuffleTrainOrderList(trainOrder);

        // Iterate through the training sets.
        for (auto& j : trainOrder)
        {
            const Matrix2d& input{trainIn[j]};
            const Matrix1d& output{trainOut[j]};

            const bool success{feedforward(input) && backpropagate(output) && optimize(learningRate)};
            if (!success) { return false; }
        }
    }
    return true;
}

// -----------------------------------------------------------------------------
const Matrix1d& Cnn::output() const noexcept
{
    const std::size_t last{myDenseLayers.size() - 1U};
    return myDenseLayers[last]->output();
}

// -----------------------------------------------------------------------------
const Matrix2d& Cnn::convOutput() const noexcept
{
    const std::size_t last{myConvLayers.size() - 1U};
    return myConvLayers[last]->output();
}

// -----------------------------------------------------------------------------
std::size_t Cnn::convOutputSize() const noexcept
{
    const std::size_t last{myConvLayers.size() - 1U};
    return myConvLayers[last]->outputSize();
}

// -----------------------------------------------------------------------------
bool Cnn::feedforward(const Matrix2d& input) noexcept
{
    bool success{true};

    // Run feedforward operation in the convolutional layers, return false on failure.
    {
        success = myConvLayers[0U]->feedforward(input);
    
        for (std::size_t i{1U}; i < myConvLayers.size(); ++i)
        {
            auto& prevLayer{*(myConvLayers[i - 1U])};
            success &= myConvLayers[i]->feedforward(prevLayer.output());
        }
        if (!success) { return false; }
    }

    // Flatten the output from the convolutional layers, return false on failure.
    {
        success = myFlattenLayer->feedforward(convOutput());
        if (!success) { return false; }
    }

    // Run feedforward operation in the dense layers, return false on failure.
    {
        success = myDenseLayers[0U]->feedforward(myFlattenLayer->output());

        for (std::size_t i{1U}; i < myDenseLayers.size(); ++i)
        {
            auto& prevLayer{*(myDenseLayers[i - 1U])};
            success &= myDenseLayers[i]->feedforward(prevLayer.output());
        }
    }
    // Return true on success.
    return success;
}

// -----------------------------------------------------------------------------
bool Cnn::backpropagate(const Matrix1d& output) noexcept
{
    bool success{true};

    // Backpropagate through the dense layers, return false on failure.
    {
        const std::size_t last{myDenseLayers.size() - 1U};
        success = myDenseLayers[last]->backpropagate(output);
        
        for (std::size_t i{last}; i > 0U; --i)
        {
            success &= myDenseLayers[i - 1U]->backpropagate(myDenseLayers[i]->inputGradients());
        }
        if (!success) { return false; }
    }

    // Unflatten the input gradients from the first dense layer, return false on failure.
    {
        success = myFlattenLayer->backpropagate(myDenseLayers[0U]->inputGradients());
        if (!success) { return false; }
    }

    // Backpropagate through the convolutional layers, return false on failure.
    {
        const std::size_t last{myConvLayers.size() - 1U};
        success = myConvLayers[last]->backpropagate(myFlattenLayer->inputGradients());

        for (std::size_t i{last}; i > 0U; --i)
        {
            success &= myConvLayers[i - 1U]->backpropagate(myConvLayers[i]->inputGradients());
        }
    }
    // Return true on success.
    return success;
}

// -----------------------------------------------------------------------------
bool Cnn::optimize(const double learningRate) noexcept
{
    bool success{true};

    // Optimize the convolutional layers, return false on failure.
    {
        success = myConvLayers[0U]->optimize(learningRate);
    
        for (std::size_t i{1U}; i < myConvLayers.size(); ++i)
        {
            success &= myConvLayers[i]->optimize(learningRate);
        }
        if (!success) { return false; }
    }

    // Optimize the dense layers, return false on failure.
    {
        success = myDenseLayers[0U]->optimize(myFlattenLayer->output(), learningRate);

        for (std::size_t i{1U}; i < myDenseLayers.size(); ++i)
        {
            auto& prevLayer{*(myDenseLayers[i - 1U])};
            success &= myDenseLayers[i]->optimize(prevLayer.output(), learningRate);
        }
    }
    // Return true on success.
    return success;
}
} // namespace ml::cnn

File: python/ml/cnn/cnn.py
"""Convolutional neural network (CNN) implementation."""
import random

from ml import utils
from ml.act_func.types import ActFuncType
from ml.cnn.interface import ICnn
from ml.factory.interface import IFactory
from ml.types import Matrix1d, Matrix2d, Matrix3d

# pylint: disable=too-many-arguments

class _TrainOrderList:
    """Training order list implementation."""

    def __init__(self, size) -> None:
        self._data = [0] * size
        for i in range(size):
            self._data[i] = i

    def data(self) -> tuple[int]:
        """Get the train order list.
        
        Returns:
            The train order list as a tuple.
        """
        return self._data

    def shuffle(self) -> None:
        """Shuffle the training order list."""
        size = len(self._data)
        for i in range(size):
            r = random.randint(0, size - 1)
            temp = self._data[i]
            self._data[i] = self._data[r]
            self._data[r] = temp


class Cnn(ICnn):
    """Convolutional neural network (CNN) implementation."""

    def __init__(self, factory: IFactory, conv_input: int, conv_kernel: int,
                 conv_func: ActFuncType, pool_size: int, dense_output: int,
                 dense_func: ActFuncType) -> None:
        self._conv_layers   = []
        self._dense_layers  = []
        self._factory       = factory

        # Add convolutional layers.
        self._conv_layers.append(factory.conv_layer(conv_input, conv_kernel, conv_func))

        self._conv_layers.append(factory.max_pool(self._conv_output_size(), pool_size))

        # Add a flatten layer.
        self._flatten_layer = factory.flatten_layer(self._conv_output_size())

        # Add a dense layer.
        dense_input = self._flatten_layer.output_size()
        self._dense_layers.append(factory.dense_layer(dense_input, dense_output, dense_func))

    def input_size(self) -> int:
        """Get the input size of the CNN.
        
        Returns:
            The input size of the CNN.
        """
        return self._conv_layers[0].input_size()

    def output_size(self) -> int:
        """Get the output size of the CNN.
        
        Returns:
            The output size of the CNN.
        """
        last = len(self._dense_layers) - 1
        return self._dense_layers[last].output_size()

    def _output(self) -> Matrix1d:
        last = len(self._dense_layers) - 1
        return self._dense_layers[last].output()

    def predict(self, input_data: Matrix2d) -> Matrix1d:
        """Predict based on the given input.
        
        Args:
            input_data: Input for which to predict.

        Returns:
            The predicted output.
        """
        self._feedforward(input_data)
        return self._output()

    def add_dense_layer(self, output_size: int, act_func_type: ActFuncType) -> None:
        """Add dense layer.

           The input size is automatically adjusted in accordance with the previous layer.
        
        Args:
            output_size: Output size.
            act_func_type: Activation function to use.
        """
        self._dense_layers.append(
            self._factory.dense_layer(self.output_size(), output_size, act_func_type))

    def train(self, train_in: Matrix3d, train_out: Matrix2d, epoch_count: int,
              learning_rate: float) -> bool:
        """Train the network.
        
        Args:
           train_in: Training input sets.
           train_out: Training output sets.
           epoch_count: Number of epochs to train the model.
           learning_rate: Learning rate to use during training.

           Returns:
               True on success, false on failure.
        """
        op = "training of CNN"
        # Check the input parameters, return false on failure.
        if not utils.check_learning_rate(learning_rate, op):
            return False
        if not utils.check_epoch_count(epoch_count, op):
            return False
        set_count = min(len(train_in), len(train_out))
        if not utils.check_train_set_count(set_count, op):
            return False
        # Create a training order list.
        train_order = _TrainOrderList(set_count)
        # Train the network the specified number of epochs.
        for _ in range(epoch_count):
            # Shuffle the training order list at the start of each epoch.
            train_order.shuffle()
            # Iterate through the training sets, return false on failure.
            for i in train_order.data():
                input_data  = train_in[i]
                output = train_out[i]
                success = (self._feedforward(input_data) and self._backpropagate(output)
                           and self._optimize(learning_rate))
                if not success:
                    return False
        # Return true on success.
        return True

    def _conv_output_size(self) -> int:
        last = len(self._conv_layers) - 1
        return self._conv_layers[last].output_size()

    def _conv_output(self) -> Matrix2d:
        last = len(self._conv_layers) - 1
        return self._conv_layers[last].output()

    def _feedforward(self, input_data) -> bool:
        # Run feedforward operation in the convolutional layers, return false on failure.
        success = self._conv_layers[0].feedforward(input_data)
        for i in range(1, len(self._conv_layers)):
            prev_layer = self._conv_layers[i - 1]
            success &= self._conv_layers[i].feedforward(prev_layer.output())
        if not success:
            return False
        # Flatten the output from the convolutional layers, return false on failure.
        success = self._flatten_layer.feedforward(self._conv_output())
        if not success:
            return False
        # Run feedforward operation in the dense layers.
        success = self._dense_layers[0].feedforward(self._flatten_layer.output())
        for i in range(1, len(self._dense_layers)):
            prev_layer = self._dense_layers[i - 1]
            success &= self._dense_layers[i].feedforward(prev_layer.output())
        # Return true on success.
        return success

    def _backpropagate(self, output: Matrix1d) -> bool:
        # Backpropagate through the dense layers, return false on failure.
        last = len(self._dense_layers) - 1
        success = self._dense_layers[last].backpropagate(output)
        for i in range(last, 0, -1):
            input_gradients = self._dense_layers[i].input_gradients()
            success &= self._dense_layers[i - 1].backpropagate(input_gradients)
        if not success:
            return False
        # Unflatten the input gradients from the first dense layer, return false on failure.
        input_gradients = self._dense_layers[0].input_gradients()
        success = self._flatten_layer.backpropagate(input_gradients)
        if not success:
            return False
        # Backpropagate through the convolutional layers.
        last = len(self._conv_layers) - 1
        input_gradients = self._flatten_layer.input_gradients()
        success = self._conv_layers[last].backpropagate(input_gradients)
        for i in range(last, 0, -1):
            input_gradients = self._conv_layers[i].input_gradients()
            success &= self._conv_layers[i - 1].backpropagate(input_gradients)
        # Return true on success.
        return success

    def _optimize(self, learning_rate: float) -> bool:
        # Optimize the convolutional layers, return false on failure.
        success = self._conv_layers[0].optimize(learning_rate)
        for i in range(1, len(self._conv_layers)):
            success &= self._conv_layers[i].optimize(learning_rate)
        if not success:
            return False
        # Optimize the dense layers, return false on failure.
        success = self._dense_layers[0].optimize(self._flatten_layer.output(), learning_rate)
        for i in range(1, len(self._dense_layers)):
            prev_layer = self._dense_layers[i - 1]
            success &= self._dense_layers[i].optimize(prev_layer.output(), learning_rate)
        # Return true on success.
        return success
